nexus/src/lib.rs

```diff
@@ -92,26 +96,98 @@ impl Server {
             ServerContext::new(config.deployment.rack_id, ctxlog, &config)
                 .await?;
 
-        let http_server_starter_external = dropshot::HttpServerStarter::new(
-            &config.deployment.dropshot_external,
-            external_api(),
-            Arc::clone(&apictx),
-            &log.new(o!("component" => "dropshot_external")),
-        )
-        .map_err(|error| format!("initializing external server: {}", error))?;
-
+        // Determine port choices
+
+        let (external_http_port, external_https_port, internal_http_port) =
+            match config.deployment.port_picker {
+                omicron_common::nexus_config::PortPicker::NexusChoice => {
+                    (80, 443, omicron_common::address::NEXUS_INTERNAL_PORT)
+                }
+                omicron_common::nexus_config::PortPicker::Zero => (0, 0, 0),
+            };
+
+        // Launch the internal server.
+
+        let dropshot_internal_config = dropshot::ConfigDropshot {
+            bind_address: SocketAddr::new(
+                config.deployment.internal_ip,
+                internal_http_port,
+            ),
+            request_body_max_bytes: 1048576,
+            ..Default::default()
+        };
         let http_server_starter_internal = dropshot::HttpServerStarter::new(
-            &config.deployment.dropshot_internal,
+            &dropshot_internal_config,
             internal_api(),
             Arc::clone(&apictx),
             &log.new(o!("component" => "dropshot_internal")),
         )
         .map_err(|error| format!("initializing internal server: {}", error))?;
+        let http_server_internal = http_server_starter_internal.start();
 
+        // Launch the external server(s).
+        //
+        // - The HTTP server is unconditionally started.
+        // - The HTTPS server is started if the necessary certificate files
+        // exist.
+        //
+        // TODO: Consider changing this disposition, making "HTTPS" the default,
+        // and returning an error if the certificates don't exist. Doing so
+        // would be the more secure long-term plan, but would make gradual
+        // deployment of this feature more difficult.
+
+        let cert_file = PathBuf::from("/var/nexus/certs/cert.pem");
```

Comment:
This seems unlikely to work on development machines.

Why not make this part of `dropshot_external`, with `/var/nexus/certs/cert.pem` the value in the SMF config file and something like "out/certs" the value in nexus/examples/config.toml?

Reply:
FWIW, this file is only used `if cert_file.exists() && key_file.exists()`.

I was kinda opposed to using the config-file-supplied certificate, because longer-term, it seemed like this value would be:

- Passed from RSS into Nexus: https://github.com/oxidecomputer/omicron/issues/1528 , as part of the rack initialization request 
- Stored in CRDB (and eventually a more secure key store, like Vault): https://github.com/oxidecomputer/omicron/pull/1500/files#r932591199

So although we are currently plumbing the certificates through the package system, I don't expect that to stay the case for very long.

Reply:
> FWIW, this file is only used if cert_file.exists() && key_file.exists().

What happens if the path is missing on a production system?  It seems like that should produce a fatal error.

> I was kinda opposed to using the config-file-supplied certificate

I can see where you're coming from.  At the same time, it seems like an essential developer or demo tool to be able to start Nexus with a specific certificate?

Is that a good idea for the production config?  I'm not sure.  It might depend on the expected behavior in the longer term.  Is it that Nexus reaches out to CockroachDB/Vault to get the certificate, then loads that into memory, and then starts the HTTPS server with it?  If that's true, then what do we do if Vault is down?  I could see: (1) not coming up at all until we can get the cert, (2) storing the last-known-good certificate to a file and trying to use that until we can get the cert, (3) generating a self-signed certificate and using that until we're able to access the real one.  These all have different tradeoffs.  I've been assuming Nexus could come up (albeit not very usefully) even if everything else in the world was down so that we could at least ask it why it wasn't able to come up.  Anyway, if we go with (2) or (3), then it seems like accepting a file path in the config file would be a reasonable stepping stone.

(Along these lines, I guess hardcoded, deployment-specific absolute paths in source files seem like a code smell to me, even if we gracefully handle them being missing.)